dev_guide/builds.adoc

```diff
@@ -846,20 +846,28 @@ source:
 
 == Binary Source
 
-When the `*BuildConfig.spec.source.type*` is `*Binary*`, the build will expect a
-binary as input, and an inline Dockerfile is optional.
-
-The binary is generally assumed to be a tar, gzipped tar, or zip file depending
-on the strategy. For `*Docker*` builds, this is the build context and an
-optional Dockerfile may be specified to override any Dockerfile in the build
-context. For `*Source*` builds, this is assumed to be an archive as described
-above. For `*Source*` and `*Docker*` builds, if `*binary.asFile*` is set the
-build context will consist of a single file named by the value of `*binary.asFile*`. 
-The `*contextDir*` field may be used when an archive is provided. Custom builds will 
-receive this binary as input on standard input (`stdin`).
-
-A binary source potentially extracts content (if `*asFile*` is not set), in which case `*contextDir*`
-allows changing to a subdirectory within the content before the build executes.
+When the `*BuildConfig.spec.source.type*` is `*Binary*`, the source content will be sent
+as input to the build via a binary stream and  `oc start-build` should be executed with either
+the `--from-dir`, `--from-file`, or `--from-repo` options.
+
+The binary stream will be processed in one of two ways:
+
+1) If `*binary.asFile*` is set, then the bytes from the binary stream will be considered a single file.  The binary
+stream will be stored in a file on the builder's file system with the name as specified as `*binary.asFile*` for the builder to use. 
```

Comment:
OK ... I thought I would look at some of the tests to confirm the interpretation of the behavior.

I found https://github.com/openshift/origin/blob/master/test/extended/builds/start.go#L85-L146

Turns out, you can use --from-dir/from-file/from-repo even when you do *NOT* have a binary source type in the BC.  Circling back and going through the `oc start-build` code path, using --from* results in a BinaryBuildRequest, and by the time it lands on the receiving side, that object is translated into a BinaryBuildSource (a la the source type Binary) ... so the --from* options are a dynamic way to turn on binary builds.

So I wondered about the source.type field in general, and discovered that it is removed in https://github.com/openshift/origin/blob/master/pkg/build/api/types.go, but still exists in the older v1 version, https://github.com/openshift/origin/blob/master/pkg/build/api/v1/types.go.

So I revisited https://github.com/openshift/origin/blob/master/pkg/build/builder/source.go#L44-L101, and sure enough, it in fact checks all the different *types* (binary, git, image, dockerfile) and essentially will overlay them from binary, to git, to images, to dockerfile as part of retrieving the *source*.

So  this implies to me that post v1, you can have multiple source types in a build config (if true, and if that was well known, it has previously escaped my awareness).

So ... my updated take on the net for the doc:
0) not with this PR, but if the removal of build.source.type becomes the next version (and is consumed by etcd I suppose), we have references to type that need to be removed.
1) If type binary.asFile is specified in the build config, and none of the --from* options are used with start-build, then that file must exist in the builder image
2) Otherwise, --from-file must be specified on the start-build, and the data comes over stdin
3) a binary build type without an as-file (that field is marked omitemtpy), and without the start-build having a --from* option, means either the source is already baked in the builder image (is that at all an interesting scenario ??), or it is totally not useful.
4) so there is essentially a matrix of behavior permutations with the bc source.binary obj and the `oc start-build --from-*` options ... 

I'll go ahead and start updating the PR based on my above interpretation, but of course will adjust as further responses occur.